spec/unit/knife/cookbook_metadata_spec.rb

```diff
@@ -174,6 +175,15 @@
         expect(@stderr.string).to match /#{description} blah/im
       end
     end
+    it "should raise a syntax error on invalid JSON" do
+      expect(File).to receive(:exist?).with("#{@cookbook_dir}/foobar/metadata.json").
+        and_return(true)
+      expect(IO).to receive(:read).with("#{@cookbook_dir}/foobar/metadata.json").
+        and_return(@invalid_json_data)
```

Comment:
i really dislike stubbing the filesystem like this.   if you try to read the file provider unit tests you'll discover where this madness winds up.

its generally preferable to consider the filesystem to be part of the "system-under-test" (SUT) and to just create a real file.  either dynamically generate a real tempfile, or use a cookbook fixture which has a real metadata.rb file in it with bad data.

that violates the principle that "unit tests should only test the object in question" and expanding the unit test to cover other objects (like the filesystem on the whole) but the amount of collaboration between objects and the underlying filesystem is huge -- a POSIX filesystem is an extremely flawed API from the perspective of testing, and just gets worse when you layer all of ruby's file classes on top of it (FileUtils, File, IO, Pathname, etc, etc) and going down the route of a MockFS or FakeFS style mocking of the entire filesystem API is pretty much madness.

it could be argued that the resultant tests using real files aren't strictly "unit" tests, but that gets into the philosophy of unit testing -- and we have lots of "unit" testing that expands the SUT to encompass other objects.  when i was learning about unit testing, i tried to 'fix' them and it wound up extremely messy and brittle.   people could argue that is because the design of the code should be bad and it should all be rewritten to have less collaboration between objects and simpler interfaces, and i'd totally agree, but lacking the time to rewrite the codebase from scratch its better just to be flexible in the definition of unit testing.

so, best practice i'd say would be to almost never stub filesystem calls.

of course we stub filesystem calls all over the place in tests, but there's a million things in our code that are horrible examples and need cleaning up...